### ISSUE TITLE
API compatibility at 1.4.8-4/5

### ISSUE THREAD

**Issue by dongjoon-hyun** [luben/zstd-jni#161]
ZSTD JNI 1.4.8-4 and 1.4.8-5 has a breaking change like the following.

```
Tests in error: 
testToAvroCommandWithZstdCompression(org.apache.parquet.cli.commands.ToAvroCommandTest):
com.github.luben.zstd.ZstdOutputStream.setCloseFrameOnFlush(Z)Lcom/github/luben/zstd/ZstdOutputStream;
```

Currently, `Apache Spark`/`Apache Parquet`/`Apache Avro` master branches are using 1.4.8. However, the released version depends on old APIs. The problem of downstream is that we cannot upgrade all projects at the same time.

First, we cannot upgrade `Apache Parquet` project because the existing Avro-related UT fails.
- https://github.com/apache/parquet-mr/pull/874 (PARQUET-1988: Upgrade to ZSTD 1.4.8-5)

Second, this breaking change blocks `Apache Spark` community from adding `Zstandard` compression to `Avro` data source.
- https://issues.apache.org/jira/browse/SPARK-34479 (Add zstandard codec to spark.sql.avro.compression.codec)

Could you help the downstream use this latest ZSTD JNI, please?

**Comment by dongjoon-hyun** [comment]
cc @wangyum , @dbtsai

**Comment by wangyum** [comment]
For Avro side. It seems zstd [1.4.5-12](https://github.com/apache/avro/blob/release-1.10.1/lang/java/pom.xml#L64) is not compatible with [1.4.8-4](https://github.com/apache/spark/blob/331c6fd4efcb337d903b7179b05997dca2dae2a8/pom.xml#L703)/5.

**Comment by luben** [comment]
Hi,

In 1.4.8-5 I tried to restore the API compatibility with versions before 1.4.8-4 but may be I am missing something. The issue with 1.4.8-4 was that methods like `setCloseFrameOnFlush` will return object of the base class. With 1.4.8-5 this should not be the case.

The changes account to creating a package local base classes for Input/Output streams and making the old classes inherit from them, also adding new *NoFinalizer variants that don't override the `finalize` method. There is also a creative use of generics so that the methods defined in the base class return object of the child class.

@wangyum, @dongjoon-hyun, can I get some examples how it breaks the compatibility? I am not sure what expectations are broken.

If the introduced inheritance is the problem I can refactor it to use composition/delegation.

**Comment by luben** [comment]
I will try to make a version during the weekend that uses the composition approach - I think it should restore all compatibility with the previous API. Today I got pinged about breaking Scala-2.11 due to https://github.com/scala/bug/issues/7936

**Comment by dongjoon-hyun** [comment]
Thank you so much for your attention and direction, @luben . Ya, in Apache Spark community, I used 1.4.8-4 and then quickly upgrade to bring the change for `setCloseFrameOnFlush`. Apache Spark itself works because it uses ZSTD JNI directly so far.

The error message was copied from `GitHub Action log` from the my first link (apache/parquet-mr#874 PARQUET-1988: Upgrade to ZSTD 1.4.8-5). When I tried to upgrade Apache Parquet from 1.4.8-3 to 1.4.8-5, the unit test about Avro fails still with the above error message. I'm still looking at that because I'm not familiar with Avro internal.
- https://github.com/apache/parquet-mr/runs/1965195510

For @wangyum 's comment, I didn't try it yet, But, he is trying to use Avro 1.10.x library's ZSTD feature but it's blocked because Apache Spark `master` branch already goes with `1.4.8-5`.

For the Scala version, we are currently using Scala 2.12.10. I'll also take a look at that Scala bug you shared. Thank you again. I'll also try to provide more direct example.

**Comment by dongjoon-hyun** [comment]
BTW, To @wangyum .
I tested `Avro` branch-1.10 with the new zstd 1.4.8-5. It passed all the Avro tests. So, it's not source-level incompatibility, right?
> For Avro side. It seems zstd 1.4.5-12 is not compatible with 1.4.8-4/5.
```
-    <zstd-jni.version>1.4.5-12</zstd-jni.version>
+    <zstd-jni.version>1.4.8-5</zstd-jni.version>
```

**Comment by luben** [comment]
Yes, it should be source level compatible. I am going forward with the refactoring as there may me some incompatibilities at runtime (apart from the Scala bug that was fixed in 2.12.3)

**Comment by dongjoon-hyun** [comment]
Thank you. For `Apache Parquet` UT failure, the following is the detail.
- Java language
- `ZSTD JNI 1.4.8-5` and `AVRO 1.10.1` are used together.
- `NoSuchMethodError` StackTrace
```
java.lang.NoSuchMethodError: com.github.luben.zstd.ZstdOutputStream.setCloseFrameOnFlush(Z)Lcom/github/luben/zstd/ZstdOutputStream;

	at org.apache.avro.file.ZstandardLoader.output(ZstandardLoader.java:40)
	at org.apache.avro.file.ZstandardCodec.compress(ZstandardCodec.java:67)
	at org.apache.avro.file.DataFileStream$DataBlock.compressUsing(DataFileStream.java:386)
	at org.apache.avro.file.DataFileWriter.writeBlock(DataFileWriter.java:407)
	at org.apache.avro.file.DataFileWriter.sync(DataFileWriter.java:428)
	at org.apache.avro.file.DataFileWriter.flush(DataFileWriter.java:437)
	at org.apache.avro.file.DataFileWriter.close(DataFileWriter.java:460)
	at org.apache.parquet.cli.commands.ToAvroCommand.run(ToAvroCommand.java:110)
	at org.apache.parquet.cli.commands.AvroFileTest.toAvro(AvroFileTest.java:49)
	at org.apache.parquet.cli.commands.AvroFileTest.toAvro(AvroFileTest.java:35)
	at org.apache.parquet.cli.commands.ToAvroCommandTest.testToAvroCommandWithZstdCompression(ToAvroCommandTest.java:89)
```

**Comment by luben** [comment]
I am preparing a release with the latest commit that I hope will restore the runtime compatibility also.

**Comment by dongjoon-hyun** [comment]
Great! I'll reopen my Apache Parquet PR when the new release is out and share the result here.

**Comment by dongjoon-hyun** [comment]
I verified locally and it works.
- https://github.com/apache/parquet-mr/pull/874 (Reopened for Apache Parquet)
- https://github.com/apache/spark/pull/31674 (Opened for Apache Spark to unblock Avro issue)

**Comment by dongjoon-hyun** [comment]
Thank you so much, @luben . I'll close this issue.

**Comment by iemejia** [comment]
Avro side merged and getting released 'soon' as part of 1.10.2 release. Great to see cross project collaboration working so well. Thanks @luben and @dongjoon-hyun !
